File: lab3 Java 21/src/MarkerController.java
// MarkerController.java
//
// lab3 Java 21
//
// created by Anhelina Mankewich, 2 course, group 12

package src;

import java.util.ArrayList;
import java.util.List;
import java.util.Random;
import java.util.concurrent.atomic.AtomicBoolean;
import java.util.concurrent.locks.Condition;
import java.util.concurrent.locks.ReentrantLock;

public class MarkerController {
    private final MarkerArray array;
    private final List<MarkerInfo> markers;
    private final ReentrantLock ctrlLock;
    private final Condition ctrlCondition;
    private final AtomicBoolean started;
    private final ReentrantLock startLock;
    private final Condition startCondition;

    private static class MarkerInfo {
        Thread thread;
        AtomicBoolean waiting;
        AtomicBoolean terminate;
        AtomicBoolean alive;
        ReentrantLock cvLock;
        Condition condition;

        MarkerInfo() {
            this.waiting = new AtomicBoolean(false);
            this.terminate = new AtomicBoolean(false);
            this.alive = new AtomicBoolean(true);
            this.cvLock = new ReentrantLock();
            this.condition = this.cvLock.newCondition();
        }
    }

    public MarkerController(int arraySize) {
        if (arraySize <= 0) {
            throw new IllegalArgumentException("Array size must be positive");
        }
        this.array = new MarkerArray(arraySize);
        this.markers = new ArrayList<>();
        this.ctrlLock = new ReentrantLock();
        this.ctrlCondition = ctrlLock.newCondition();
        this.started = new AtomicBoolean(false);
        this.startLock = new ReentrantLock();
        this.startCondition = startLock.newCondition();
    }

    public void startMarkers(int count) {
        if (count <= 0) return;

        ctrlLock.lock();
        try {
            markers.clear();
            for (int i = 0; i < count; i++) {
                markers.add(new MarkerInfo());
            }

            for (int i = 0; i < count; i++) {
                final int id = i + 1;
                MarkerInfo info = markers.get(i);
                info.thread = new Thread(() -> markerWorker(id));
                info.thread.start();
            }
        } finally {
            ctrlLock.unlock();
        }
    }

    public void signalStartAll() {
        startLock.lock();
        try {
            started.set(true);
            startCondition.signalAll();
        } finally {
            startLock.unlock();
        }
    }

    private void markerWorker(int id) {
        int arraySize = array.size();
        Random random = new Random(id);

        startLock.lock();
        try {
            while (!started.get()) {
                startCondition.await();
            }
        } catch (InterruptedException e) {
            Thread.currentThread().interrupt();
            return;
        } finally {
            startLock.unlock();
        }

        while (true) {
            ctrlLock.lock();
            try {
                if (markers.get(id - 1).terminate.get()) {
                    break;
                }
            } finally {
                ctrlLock.unlock();
            }

            int index = random.nextInt(arraySize);

            if (array.get(index) == 0) {
                sleepMs(5);
                if (array.tryMark(index, id)) {
                    sleepMs(5);
                    continue;
                }
            }

            int markedCount = array.countMarkedBy(id);

            synchronized (System.out) {
                System.out.println("Marker " + id + " | Marked: " + markedCount + " | Blocked at index: " + index);
            }

            markers.get(id - 1).waiting.set(true);
            
            ctrlLock.lock();
            try {
                ctrlCondition.signal();
            } finally {
                ctrlLock.unlock();
            }

            MarkerInfo info = markers.get(id - 1);
            info.cvLock.lock();
            try {
                while (!info.terminate.get() && info.waiting.get()) {
                    info.condition.await();
                }
            } catch (InterruptedException e) {
                Thread.currentThread().interrupt();
                break;
            } finally {
                info.cvLock.unlock();
            }

            if (markers.get(id - 1).terminate.get()) {
                break;
            }
        }

        array.clearMarks(id);
        markers.get(id - 1).alive.set(false);
        
        ctrlLock.lock();
        try {
            ctrlCondition.signal();
        } finally {
            ctrlLock.unlock();
        }

        synchronized (System.out) {
            System.out.println("Marker " + id + " terminated and cleared its marks");
        }
    }

    public void waitAllBlocked() {
        ctrlLock.lock();
        try {
            while (true) {
                int active = 0;
                int waiting = 0;
                for (MarkerInfo marker : markers) {
                    if (marker.alive.get()) {
                        active++;
                        if (marker.waiting.get()) {
                            waiting++;
                        }
                    }
                }
                if (active > 0 && waiting == active) {
                    break;
                }
                ctrlCondition.await();
            }
        } catch (InterruptedException e) {
            Thread.currentThread().interrupt();
        } finally {
            ctrlLock.unlock();
        }
    }

    public boolean terminateMarker(int markerId) {
        if (markerId < 1 || markerId > markers.size()) {
            return false;
        }

        ctrlLock.lock();
        try {
            if (!markers.get(markerId - 1).alive.get()) {
                return false;
            }
            markers.get(markerId - 1).terminate.set(true);
            
            MarkerInfo info = markers.get(markerId - 1);
            info.cvLock.lock();
            try {
                info.condition.signal();
            } finally {
                info.cvLock.unlock();
            }
        } finally {
            ctrlLock.unlock();
        }

        try {
            markers.get(markerId - 1).thread.join();
        } catch (InterruptedException e) {
            Thread.currentThread().interrupt();
            return false;
        }

        return true;
    }

    public void resumeAll() {
        ctrlLock.lock();
        try {
            for (MarkerInfo marker : markers) {
                if (marker.waiting.get() && marker.alive.get() && !marker.terminate.get()) {
                    marker.waiting.set(false);
                    marker.cvLock.lock();
                    try {
                        marker.condition.signal();
                    } finally {
                        marker.cvLock.unlock();
                    }
                }
            }
        } finally {
            ctrlLock.unlock();
        }
    }

    public void printArray() {
        int[] snapshot = array.copy();
        for (int i = 0; i < snapshot.length; i++) {
            System.out.print(snapshot[i]);
            if (i + 1 < snapshot.length) {
                System.out.print(" ");
            }
        }
        System.out.println();
    }

    public boolean hasActiveMarkers() {
        for (MarkerInfo marker : markers) {
            if (marker.alive.get()) {
                return true;
            }
        }
        return false;
    }

    public int aliveCount() {
        int count = 0;
        for (MarkerInfo marker : markers) {
            if (marker.alive.get()) {
                count++;
            }
        }
        return count;
    }

    public int findFirstAlive() {
        for (int i = 0; i < markers.size(); i++) {
            if (markers.get(i).alive.get()) {
                return i + 1;
            }
        }
        return 0;
    }

    public void joinAll() {
        ctrlLock.lock();
        try {
            for (MarkerInfo marker : markers) {
                if (marker == null) continue;
                marker.terminate.set(true);
                marker.cvLock.lock();
                try {
                    marker.condition.signal();
                } finally {
                    marker.cvLock.unlock();
                }
            }
        } finally {
            ctrlLock.unlock();
        }

        for (MarkerInfo marker : markers) {
            if (marker != null && marker.thread != null) {
                try {
                    marker.thread.join();
                } catch (InterruptedException e) {
                    Thread.currentThread().interrupt();
                }
            }
        }
        markers.clear();
    }

    private void sleepMs(int ms) {
        try {
            Thread.sleep(ms);
        } catch (InterruptedException e) {
            Thread.currentThread().interrupt();
        }
    }
}

File: lab3 C++17/src/marker_controller.cpp
// marker_controller.cpp
//
// lab3 C++17
//
// created by Anhelina Mankewich, 2 course, group 12

#include "marker_controller.h"
#include <iostream>
#include <chrono>
#include <thread>
#include <cstdlib>
#include <ctime>

std::mutex MarkerController::cout_mtx;

MarkerController::MarkerController(std::size_t array_size) {
    if (array_size == 0) {
        throw std::invalid_argument("array_size must be > 0");
    }
    array = std::make_shared<MarkerArray>(array_size);
}

MarkerController::~MarkerController() {
    join_all();
}

void MarkerController::sleep_ms(int ms) {
    std::this_thread::sleep_for(std::chrono::milliseconds(ms));
}

void MarkerController::start_markers(std::size_t count) {
    if (count == 0) return;
    
    std::lock_guard<std::mutex> lock(ctrl_mtx);
    markers.clear();
    markers.reserve(count);
    
    for (std::size_t i = 0; i < count; ++i) {
        markers.push_back(std::make_unique<MarkerInfo>());
    }
    
    for (std::size_t i = 0; i < count; ++i) {
        markers[i]->thread = std::thread(&MarkerController::marker_worker, this, i + 1);
    }
}

void MarkerController::signal_start_all() {
    {
        std::lock_guard<std::mutex> lock(start_mtx);
        started.store(true);
    }
    start_cv.notify_all();
}

void MarkerController::marker_worker(std::size_t id) {
    std::size_t array_size = array->size();
    
    std::srand(static_cast<unsigned int>(id));
    
    {
        std::unique_lock<std::mutex> lock(start_mtx);
        start_cv.wait(lock, [this] { return started.load(); });
    }
    
    while (true) {
        {
            std::lock_guard<std::mutex> lock(ctrl_mtx);
            if (markers[id - 1]->terminate.load()) break;
        }
        
        std::size_t index = static_cast<std::size_t>(std::rand()) % array_size;
        
        if (array->get(index) == 0) {
            sleep_ms(5);
            if (array->try_mark(index, static_cast<int>(id))) {
                sleep_ms(5);
                continue;
            }
        }
        
        std::size_t marked_count = array->count_marked_by(static_cast<int>(id));
        
        {
            std::lock_guard<std::mutex> lock(cout_mtx);
            std::cout << "Marker " << id 
                      << " | Marked: " << marked_count 
                      << " | Blocked at index: " << index << std::endl;
        }
        
        markers[id - 1]->waiting.store(true);
        ctrl_cv.notify_one();
        
        std::unique_lock<std::mutex> lock(markers[id - 1]->cv_mtx);
        markers[id - 1]->cv.wait(lock, [this, id] {
            return markers[id - 1]->terminate.load() || !markers[id - 1]->waiting.load();
        });
        
        if (markers[id - 1]->terminate.load()) break;
    }
    
    array->clear_marks(static_cast<int>(id));
    markers[id - 1]->alive.store(false);
    ctrl_cv.notify_one();
    
    {
        std::lock_guard<std::mutex> lock(cout_mtx);
        std::cout << "Marker " << id << " terminated and cleared its marks" << std::endl;
    }
}

void MarkerController::wait_all_blocked() {
    std::unique_lock<std::mutex> lock(ctrl_mtx);
    ctrl_cv.wait(lock, [this] {
        std::size_t active = 0;
        std::size_t waiting = 0;
        for (auto& marker : markers) {
            if (marker->alive.load()) {
                ++active;
                if (marker->waiting.load()) ++waiting;
            }
        }
        return active > 0 && waiting == active;
    });
}

bool MarkerController::terminate_marker(std::size_t marker_id) {
    if (marker_id == 0 || marker_id > markers.size()) return false;
    
    {
        std::lock_guard<std::mutex> lock(ctrl_mtx);
        if (!markers[marker_id - 1]->alive.load()) return false;
        markers[marker_id - 1]->terminate.store(true);
        markers[marker_id - 1]->cv.notify_one();
    }
    
    if (markers[marker_id - 1]->thread.joinable()) {
        markers[marker_id - 1]->thread.join();
    }
    
    return true;
}

void MarkerController::resume_all() {
    std::lock_guard<std::mutex> lock(ctrl_mtx);
    for (auto& marker : markers) {
        if (marker->waiting.load() && marker->alive.load() && !marker->terminate.load()) {
            marker->waiting.store(false);
            marker->cv.notify_one();
        }
    }
}

void MarkerController::print_array(std::ostream& os) const {
    std::vector<int> snapshot = array->copy();
    for (std::size_t i = 0; i < snapshot.size(); ++i) {
        os << snapshot[i];
        if (i + 1 < snapshot.size()) os << " ";
    }
    os << std::endl;
}

bool MarkerController::has_active_markers() const {
    for (auto& marker : markers) {
        if (marker->alive.load()) return true;
    }
    return false;
}

std::size_t MarkerController::alive_count() const {
    std::size_t count = 0;
    for (auto& marker : markers) {
        if (marker->alive.load()) ++count;
    }
    return count;
}

std::size_t MarkerController::find_first_alive() const {
    for (std::size_t i = 0; i < markers.size(); ++i) {
        if (markers[i]->alive.load()) return i + 1;
    }
    return 0;
}

void MarkerController::join_all() {
    std::lock_guard<std::mutex> lock(ctrl_mtx);
    for (auto& marker : markers) {
        if (!marker) continue;
        marker->terminate.store(true);
        marker->cv.notify_one();
    }
    for (auto& marker : markers) {
        if (marker && marker->thread.joinable()) {
            marker->thread.join();
        }
    }
    markers.clear();
}